### ISSUE TITLE
`CatTransform` should work with `event_dim > 0` transforms

### ISSUE THREAD

**Issue by vikigenius** [pytorch/pytorch#44530]
## 🐛 Bug

Just opening a bug report corresponding to attempted pull request: https://github.com/pytorch/pytorch/pull/36878

As described the issue is that CatTransform does not work for event_dim > 0.

## To Reproduce

```py
import torch
from torch.distributions import Normal
from torch.distributions.transforms import AffineTransform, CatTransform

prior = Normal(torch.zeros(16, 8), torch.ones(16, 8))
t1 = AffineTransform(0, 1, event_dim=1)
t2 = AffineTransform(0, 1, event_dim=1)

tc = CatTransform([t1, t2], dim=1, lengths=[4, 4])

ps = prior.sample()
tps = tc(ps)  # Works fine, returns a tensor of size ([16, 8])
lad = tc.log_abs_det_jacobian(ps, tps)  # This line throws an error:
```
```
Traceback (most recent call last):
  File "/home/void/scripts/test_pytorch.py", line 14, in <module>
    lad = tc.log_abs_det_jacobian(ps, tps)  # This line throws an error:
  File "/home/void/.miniconda3/envs/textvinf/lib/python3.8/site-packages/torch/distributions/transforms.py", line 693, in log_abs_det_jacobian
    return torch.cat(logdetjacs, dim=self.dim)
IndexError: Dimension out of range (expected to be in range of [-1, 0], but got 1)
```

## Expected behavior
log_abs_det_jacobian should work with CatTransform even for event_dim > 0

cc @vincentqb @fritzo @neerajprad @alicanb @vishwakftw

**Comment by fritzo** [comment]
@vikigenius I believe the issue here is that `CatTransform`'s `dim` argument refers to a batch dimension (since we can't cat along an event dimension), but your data `ps` has batch shape `(16,)` which is too narrow for `dim=1`. You should be able to either change `dim=0` to cat along the 16-long dimension, or expand your batch shape to `(1, 16)` so that `dim=1` refers to the 16-long dimension.

I realize the docs are brief and possibly confusing. We'd appreciate any PR clarifying the docs in a way you think would confuse fewer people. Thanks for the report!

**Comment by vikigenius** [comment]
@fritzo but we are actually able to cat along an event dimension right? That's why I got an output tensor of shape (16, 8). It's just that the log_det_jacobian errors out. The discussion in the PR is pretty on point I think. The log_det_jacobian already sums out the event_dim and thus you cannot take into account the concatenation.

What I am looking for is basically the equivalent of Tensorflow's blockwise bijector: https://www.tensorflow.org/probability/api_docs/python/tfp/bijectors/Blockwise

**Comment by fritzo** [comment]
@vikigenius hmm I think I see your point now, and I think my previous comment was wrong. Does this look right to you:
1. The concatenation `dim` can be either a batch dim or an event dim.
2. If `dim` is a batch dim then the current logic is correct, i.e. concatenate the log-det-jacobians:
    ```py
    logdeghacs = []
    for trans, length in zip(self.transforms, self.lengths):
        ...
        logdetjacs.append(trans.log_abs_det_jacobian(...))
    return torch.cat(logdetjacs, dim=self.dim)
    ```
3. If `dim` is an event dim then we need new logic to instead *add* the component log-det-jacobians:
    ```py
    ...
    return sum(logdetjacs)
    ```

If this sounds correct, then I think there's a pretty simple fix:

<detail>

```diff
diff --git a/torch/distributions/transforms.py b/torch/distributions/transforms.py
index 09e00d55e8..eddce21c25 100644
--- a/torch/distributions/transforms.py
+++ b/torch/distributions/transforms.py
@@ -633,6 +633,7 @@ class CatTransform(Transform):
     """
     def __init__(self, tseq, dim=0, lengths=None, cache_size=0):
         assert all(isinstance(t, Transform) for t in tseq)
+        assert len({t.event_dim for t in tseq}) == 1
         if cache_size:
             tseq = [t.with_cache(cache_size) for t in tseq]
         super(CatTransform, self).__init__(cache_size=cache_size)
@@ -686,7 +687,15 @@ class CatTransform(Transform):
             yslice = y.narrow(self.dim, start, length)
             logdetjacs.append(trans.log_abs_det_jacobian(xslice, yslice))
             start = start + length  # avoid += for jit compat
-        return torch.cat(logdetjacs, dim=self.dim)
+        dim = self.dim if self.dim < 0 else self.dim - x.dim()
+        if dim < 0:  # concatenate along a batch_dim
+            return torch.cat(logdetjacs, dim=dim)
+        else:  # concatenate along an event_dim
+            return sum(logdetjacs)
+
+    @property
+    def event_dim(self):
+        return self.transforms[0].event_dim

     @property
     def bijective(self):
```

</details>

**Comment by fritzo** [comment]
Oops ha ha looks like we came up with the same solution last April, I don't know why it got stuck in review...

**Comment by vikigenius** [comment]
@fritzo , it does look correct, although thinking about the discussion in the pull request logically, why would we even need to worry about the concatenation ever the batch dimensions? A probability transform should operate only along the event dimension right? 

Concatenating two distributions on the batch dim doesn't even make sense. The user can simply concatenate the sampled vectors and the log-jacobians of the individual transformation themselves. The resulting sample is not even a sample from a single distribution it is basically a vector of samples from different distributions. I know this may actually be a breaking change, but I don't think we should allow concatenation along the batch dimension at all. That's not why the Transform class exists.

Tensorflow's blockwise bijector completely bypasses this problem by not giving the user the choice of the dimension at all, and assumes that no transform has an event_dim > 1 which basically means you can assume a target event_dim of 1, because concatenation is automatically a vector transform along the event dimension.

**Comment by fritzo** [comment]
cc @ahmadsalim who created `CatTransform` and `StackTransform` in #11868 resolving #11755

**Comment by fritzo** [comment]
cc @stefanwebb who maintains Pyro's transforms library. What are stack and cat used for?

**Comment by ahmadsalim** [comment]
My original idea was that `CatTransform` and `StackTransform` would allow applying component-wise transformations for multi-variate distributions.

Concrete example:
* I have a multi-variate Normal variable X and I want to apply a circular transform to the first n dimensions of it.

I have to agree that it is primarily meant for `event_dim`, but it has been a while and I cannot remember the concrete details of implementation. Maybe copying the behavior of blockwise bijector from Tensorflow is closer to what we want.

**Comment by vikigenius** [comment]
Yes, exactly logically concatenating along the batch_dim doesn't even make sense. Of course you can do it, but you don't need a transform for that. Just concatenate the different samples together.

Let me give a simple example:
Say you have a random variable X ~ Normal
Then Y = exp(X) would be Log Normal

Defining two transforms [Exp(), Identity()] and using Cat Transform of these on X along the event dimension would give you a random variable with some joint distribution (Y, X) where Y is Log Normal and X is Normal.

Concatenating along the Batch Dimension is not even a distribution it is just a collection of samples where the first B samples are Log Normal and the next B samples are Normal (where B is the batch_size).

**Comment by fritzo** [comment]
@vikigenius I like your example `[Exp(), Identity()]`. However note that both of those transforms are scalar transforms: they have `event_dim=0` and can be concatenated only along batch dimensions. Note the distinction between (a) the`.event_shape` of a base distribution and (b) the `.event_dim` of a transform, say in a `TransformedDistribution(base_dist, transform)`. In @ahmadsalim's example, `base_dist.event_shape=(6,)` but `transform.event_dim=0`. The extra `cat() ->sum()` operation would currently be accomplished in `TransformedDistribution.log_prob()`:
https://github.com/pytorch/pytorch/blob/8b7ee33ee696f2746ae163f56edd931bc02908ee/torch/distributions/transformed_distribution.py#L116-L117
The new `sum()` logic in #36878 would apply only to transforms with `.event_dim > 0` such as `StickBreakingTransform` or `LowerCholeskyTransform`.

In summary, I agree that in principle it only makes sense to define `TransformedDistribution(base_dist, CatTransform(...))` where concatenation is along an event dimension, but even to construct that `CatTransform` of scalar transforms we would need either #36878 or a new `IndependentTransform`. Either or both of those solutions seems fine to me, but #36878 seems more backwards-compatible. WDYT?

**Comment by vikigenius** [comment]
@fritzo Agreed, ultimately we need to strike a balance and if people want to use it to concatenate along batch_dims, let them it's fine, they should just keep in mind that ultimately it's not going to be a joint distribution. So I guess #36878 seems to get the job done while being backwards compatible. We could just include the clarification in the docs.

**Comment by vikigenius** [comment]
@fritzo but also, if you consider my example and if someone wanted to concatenate along the event_dim, they would have to do the Indpendent thing themsemves and make sure they have event_dim = 1 before using CatTransform is that right?

**Comment by fritzo** [comment]
@vikigenius sorry I don't understand your example, could you rewrite it in terms of `torch.distributions` objects? For example `torch.distributions` does not have a notion of random variable, only distributions and transforms.

Trying to interpret your example in terms of `torch.distributions` concepts I think we could either 
```py
t = CatTransform([ExpTransform(), identity_transform], lengths=[3,3], dim=-1)
x = torch.rand(B, 3+3)
assert t(x).shape == (B, 3+3)
```
or
```py
t = CatTransform([ExpTransform(), identity_transform], lengths=[B,B], dim=-2)
x = torch.rand(B+B, 3)
assert t(x).shape == (B+B, 3)
```
but neither of those examples could apply to a "tensor x of shape [B, 3]". Note that our usage satisfies the equation
```py
CatTransform([f,g])(torch.cat([x,y])) = torch.cat([f(x), g(y)])
```

**Comment by vikigenius** [comment]
@fritzo you are right when you want to write it as a blockwise transform x should have a shape of [B, 6], my mistake. 

My example was just to illustrate mathematically why concatenating along the batch_dim doesn't really make sense generally. 

But I think it's just a confusion on my side with the terminology, the way we have transforms set up in your second example, it could technically still make sense to concatenate along your batch dimension when you apply pointwise transform to an independently distributed sample because mathematically the batch_dim is the same as the event_dim in this case because they are independent and we are doing pointwise transforms.

But I am not sure I follow your logic of branching. What happens when one transformation is supposed to have event_dim=1 and the other has event_dim=0. Now naturally it makes more sense mathematically to concatenate along the event_dim 1. What happens if someone passes 0 as the dim for CatTransorm?

**Comment by fritzo** [comment]
> What happens when one transformation is supposed to have event_dim=1 and the other has event_dim=0

It is fine to concatenate transforms of heterogeneous event_dim. For transforms, `.event_dim` means "how many rightmost dims are mixed together by this transform". So the `.event_dim` of the concatenation would be 1 since at least one component transform mixes up 1 dimension. We'll need to handle that `self.event_dim = max(t.event_dim for t in transforms)` logic in #36878, including summing the rightmost `(self.event_dim - t.event_dim)`-many dims of each transform so that the cat'd or summed log-abs-det-jacobian makes sense.

**Comment by vikigenius** [comment]
Okay now it makes perfect sense. I think your logic looks fine. Adding a few tests should make it clear that it's working as intended.

**Comment by vikigenius** [comment]
@fritzo Is there something still blocking the issue? I just came back after a short break and I see that it hasn't been fixed yet. Is there anything I can do to help?

**Comment by fritzo** [comment]
@vikigenius this issue is confusing and it looks like we pivoted multiple times. I'm closing it to avoid the confusing conversation. If you still feel there is incorrect behavior could you open up a new issue and suggest a fix, or simply open up a PR with the fix?

**Comment by vikigenius** [comment]
@fritzo The conversation was confusing, but the bug essentially is the same as provided in the issue description. I have a created a pull request that solves this. Take a look.
